src/cdef.rs

```diff
@@ -59,20 +59,57 @@ pub(crate) mod rust {
   ///
   /// # Arguments
   ///
-  /// * `elems` - A non-empty slice of integers
-  ///
-  /// # Panics
-  ///
-  /// Panics if `elems` is empty
+  /// * `elems` - A slice of 8 `i32`s
   #[inline]
-  fn first_max_element(elems: &[i32]) -> (usize, i32) {
-    // In case of a tie, the first element must be selected.
-    let (max_idx, max_value) = elems
-      .iter()
-      .enumerate()
-      .max_by_key(|&(i, v)| (v, -(i as isize)))
-      .unwrap();
-    (max_idx, *max_value)
+  fn first_max_element(
+    elems: &[i32; 8], cpu: CpuFeatureLevel,
+  ) -> (usize, i32) {
+    // Same as `first_max_element`, but implemented with AVX2 intrinsics
+    #[inline]
+    #[cfg(nasm_x86_64)]
+    #[target_feature(enable = "avx2,bmi1,bmi2")]
+    unsafe fn first_max_element_avx2(elems: &[i32; 8]) -> (usize, i32) {
+      #[cfg(target_arch = "x86")]
+      use std::arch::x86::*;
+      #[cfg(target_arch = "x86_64")]
+      use std::arch::x86_64::*;
+
+      // the compiler autovectorizes this
+      let max_val = *elems.iter().max().unwrap();
+
+      let cmp = _mm256_cmpeq_epi32(
+        _mm256_loadu_si256(elems as *const i32 as *const _),
+        _mm256_set1_epi32(max_val),
+      );
+      // this intrinsic is supposed to be for floating point, but it works
+      // fine on integer data as well
+      let mask = _mm256_movemask_ps(std::mem::transmute(cmp));
+
+      (mask.trailing_zeros() as usize, max_val)
+    }
+
+    #[inline]
+    fn _first_max_element(elems: &[i32; 8]) -> (usize, i32) {
```

Comment:
Do not use `_` it does not have the same meaning as in C.

Reply:
Thanks, fixed.